### PR TITLE
Unify metric API into the one `otel/metric` package

### PR DESCRIPTION
**Pull request by MrAlias**
Resolves: https://github.com/open-telemetry/opentelemetry-go/issues/3995

Changes:

- Flatten `metric/instrument` into `metric`
- Deprecates all previously released types from `metric/instrument` and the package itself

These changes are proposed as they resolve the following:

- The `metric/instrument` package only provides types and functionality called from the `metric` package
- The metric API is fractured across two packages currently
   - Users need to import two packages for the single task of instrumenting code with metrics
   - Users need to reference documentation from `metric/instrument` to understand the full purpose and functionality of `metric`

This proposal has been discussed before[^1][^2], why change it now?

Prior considerations were made before the `Meter` API was updated to be specification compliant. [That update](https://github.com/open-telemetry/opentelemetry-go/pull/3530) and [the following cleanup](https://github.com/open-telemetry/opentelemetry-go/pull/3575) removed the non-compliant instrument provider and the number scoped instrument packages. With those changes the instruments themselves are never meant to be created or used outside of the functionality provided by `metric`. It is a relic to keep the partition of the `metric/instrument` package when all of its use is now in `metric`.

[^1]: https://github.com/open-telemetry/opentelemetry-go/issues/2653
[^2]: https://github.com/open-telemetry/opentelemetry-go/issues/2526
[^3]: https://github.com/open-telemetry/opentelemetry-go/pull/2557#pullrequestreview-878246548

### CODE REVIEW

**Review by dmathieu — APPROVED** [review]

File: metric/doc.go [hunk 1]
```diff
@@ -26,9 +26,47 @@ instruments are created by a [Meter] which itself is created by a
...
+are used to measure the operation and performance of source code during the
+source code execution. These instruments only make measurements when the source
+code they instrument is run.
+
+All asynchronous instruments ([Int64ObservableCounter],
+[Int64ObservableUpDownCounter], [Int64ObservableGauge],
+[Float64ObservableCounter], [Float64ObservableUpDownCounter],
+[Float64ObservableGauge]) are used to measure metrics outside of the execution
+of source code. They are said to make "observations" via a callback function
+called once every measurement collection cycle.
```
Comment (pellared) [inline metric/doc.go:48]: How to add "observations" for histograms? Is it worth calling out that it is not possible?

File: metric/doc.go [hunk 2]
```diff
@@ -26,9 +26,47 @@ instruments are created by a [Meter] which itself is created by a
...
-instrument and its intended use.
+can then be built from the [Meter]'s instruments.
+
+# Instruments
+
+Each instrument is designed to make measurements of a particular type. Broadly,
+all instruments fall into two overlapping logical categories: asynchronous or
+synchronous, and int64 or float64.
+
+All synchronous instruments ([Int64Counter], [Int64UpDownCounter],
```
Comment (pellared) [inline metric/doc.go:37]: Should we tell that there is no synchronous gauge and one can use `UpDownCounter` instead?
Example metric: "room temperature goal" (in an air conditioner setting, for comparison a "actual room temperature" can be an async gauge.

**Review by pellared — APPROVED** [review]
I am aware that my comments are mostly concerns about the OTel Metrics API Spec. However, I do not think that the package users need to know/check exactly what is in the spec.

**Comment by pellared** [comment]
@seh PTAL 😉

File: metric/doc.go (hunk 1 above)
Comment (MrAlias) [reply metric/doc.go:48]: This is a copy/paste from the other package. I'd rather not edit this in this PR.

Feel free to add a PR once this merges.

File: metric/doc.go (hunk 2 above)
Comment (MrAlias) [reply metric/doc.go:37]: Same here:

> This is a copy/paste from the other package. I'd rather not edit this in this PR.
>
> Feel free to add a PR once this merges.

**Comment by seh** [comment]
Do the types and functions that had lived in the `sdk/metric/instrument` package need to be exported? It's hard to tell from the PR how the files outside this package were relying on it. In some cases it was the options that we discussed during last week's SIG meeting, and continuing to export those is still necessary.

For example, does the specification mandate that we expose types like `metric.Float64Observable`? Can we shrink the exported surface of this Go module?

**Comment by MrAlias** [comment]
> Do the types and functions that had lived in the `sdk/metric/instrument` package need to be exported? It's hard to tell from the PR how the files outside this package were relying on it. In some cases it was the options that we discussed during last week's SIG meeting, and continuing to export those is still necessary.
> 
> For example, does the specification mandate that we expose types like `metric.Float64Observable`? Can we shrink the exported surface of this Go module?

All of the instruments need to be exported as they are returned from their respective creation method of the `Meter`.

The `Observable` itself is needed to generally reference all the observable instruments that can be passed to `RegisterCallback` on the `Meter`.

**Comment by seh** [comment]
> All of the instruments need to be exported as they are returned from their respective creation method of the `Meter`.

Got it. This is not the opportunity I was hoping to find.

**Comment by MrAlias** [comment]
@Aneurysm9 thoughts?

**Comment by Aneurysm9** [comment]
> * Flatten `sdk/metric/instrument` into `sdk/metric`
> 
> * Deprecates all previously released types from `sdk/metric/instrument` and the package itself

`metric/instrument`, not `sdk/metric/instrument`, right?

Otherwise, I think this is the logical extension of the work already done to collapse the package hierarchy.

**Pull request merged**